### PR TITLE
security(dashboard): require email verification before applying profile email changes

### LINKED ISSUE #375: security(dashboard): require email verification before applying profile email changes
## Problem

\`dashboard/src/apps/auth/views/profile.rs:43-45, 88-94\` updates \`user.email\` directly when a profile update request sets a new email. There is no verification step, enabling email takeover attacks.

## Proposed fix

1. Add an \`EmailVerificationToken\` ORM model (user_id, pending_email, token_hash, expires_at, consumed_at) with migration.
2. Introduce an \`EmailSender\` trait abstraction; default impl uses \`lettre\` over SMTP. Configure via \`MAIL_FROM\` / \`MAIL_SMTP_URL\` / credentials env vars.
3. In \`profile_update\`, when the new email differs from the current value:
   - Generate a 32-byte CSPRNG token (OsRng), store its sha256 hash, send the plaintext via email link (\`/auth/verify-email/?user_id=...&token=...\`).
   - Return 202 Accepted without updating \`user.email\`.
4. Add \`GET /auth/verify-email/\` endpoint: constant-time compare, verify not expired / not consumed, atomically update \`user.email\` and mark \`consumed_at\`.
5. Delete the placeholder \"Note: ... future iteration\" comment at profile.rs:43-45.

## Scope
- \`dashboard/src/apps/auth/views/profile.rs\`
- \`dashboard/src/apps/auth/views/verify_email.rs\` (new)
- \`dashboard/src/apps/auth/models/email_verification_token.rs\` (new)
- \`dashboard/src/apps/auth/services/mailer.rs\` (new)
- \`dashboard/src/apps/auth/urls.rs\`
- Migration file for the new table

### PR DESCRIPTION
**Pull request by kent8192**
## Summary
- Add \`EmailVerificationToken\` ORM model + migration (SHA-256 hashed token, \`expires_at\`, single-use \`consumed_at\`).
- Introduce \`EmailSender\` trait with \`LettreSmtpSender\` (production) and \`NullEmailSender\` (tests); config via \`MAIL_FROM\` / \`MAIL_SMTP_URL\`.
- \`smtp://\` scheme is rejected at startup to prevent accidental cleartext transport; operators opt in to cleartext via the explicit \`smtp+insecure://\` scheme.
- SMTP URL parsing uses the \`url\` crate (RFC 3986) to correctly handle IPv6 hosts and percent-encoded credentials.
- \`profile_update\` no longer mutates \`user.email\`; it issues a 24h token (CSPRNG via \`OsRng\`) and sends a verification link, returning \`202 { "email_verification": { "status": "verification_sent" } }\`. Name changes in the same request are persisted immediately; the status code distinguishes outcomes.
- New \`POST /auth/verify-email-change/\` endpoint: the frontend confirmation page extracts the token from the URL fragment (not the query string, to avoid access-log/Referer leaks) and POSTs it as a JSON body. The endpoint hashes the submitted plaintext, looks up the hash in the DB (filtered \`consumed_at IS NULL AND expires_at > now()\`), verifies \`user_id\` binding, updates \`user.email\`, and marks the token consumed.
- The DB lookup approach is timing-safe: \`token_hash\` has a UNIQUE constraint so at most one row is returned regardless of token value, and lookup timing is dominated by the uniform index access — application-level \`ConstantTimeEq\` would add no security here.
- \`PUBLIC_BASE_URL\` being unset or empty is treated as a configuration error (returns 500) rather than silently producing broken verification links.
- Prior unconsumed tokens for a user are deleted at DB level (using the partial index on \`(user_id) WHERE consumed_at IS NULL\`) before issuing a new one.

## Test plan
- [x] \`cargo test -p reinhardt-cloud-dashboard --lib apps::auth::tests::unit\` (token primitives, NullEmailSender)
- [ ] E2E tests for \`verify_email_change\` edge cases (expired token, consumed token, user_id mismatch) — partial index and DB-level filtering are unit-testable; full request coverage deferred pending Postgres testcontainer integration

## Security notes
- Randomness: \`rand::rngs::OsRng\` (32 bytes)
- Hashing: SHA-256 (stored as hex); plaintext token only transmitted in the verification URL fragment
- Token delivery: URL fragment \`#user_id=...&token=...\` — fragments are never transmitted to the server, so they stay out of access logs and \`Referer\` headers
- Comparison: DB equality lookup by hash (UNIQUE index) — uniform timing, no application-side scan
- Token: single-use (\`consumed_at\`) + 24h expiry (\`expires_at\`) + prior unconsumed tokens deleted on reissue
- Atomicity gap: two-step update (user email then token consumed) — logged loudly; transaction plumbing is out-of-scope for this PR

Fixes #375

### CODE REVIEW

File: dashboard/src/apps/auth/views/profile.rs [hunk 1]
```diff
@@ -85,25 +91,121 @@ pub async fn profile_update(
...
 		AppError::Internal("Internal server error".to_string())
 	})?;
 
+	if let Some(pending_email) = email_change_requested {
+		issue_email_verification(&updated, &pending_email).await?;
+		let resp = serde_json::json!({ "status": "verification_sent" });
+		return Ok(Response::new(StatusCode::ACCEPTED)
+			.with_header("Content-Type", "application/json")
+			.with_body(json::to_vec(&resp)?));
+	}
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/src/apps/auth/views/profile.rs]: Addressed in commit 7be4a3e: profile_update now always returns 200 with an explicit `email_verification` status field (`null` / `sent` / `failed`) so the response truthfully reflects what was committed. Mailer-delivery failures degrade to `failed` instead of a misleading 500.

File: dashboard/src/apps/auth/views/profile.rs [hunk 2]
```diff
@@ -85,25 +91,121 @@ pub async fn profile_update(
...
+		.map_err(|e| {
+			error!("Failed to persist verification token: {e}");
+			AppError::Internal("Internal server error".to_string())
+		})?;
+
+	let base_url = std::env::var("PUBLIC_BASE_URL").unwrap_or_default();
+	let url = format!(
+		"{base_url}/auth/verify-email/?user_id={}&token={}",
+		user.id, token_plain
+	);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/src/apps/auth/views/profile.rs]: Addressed in commit 7be4a3e: missing or empty PUBLIC_BASE_URL is now treated as a configuration error (`AppError::Internal`) before any verification email is built.

File: dashboard/src/apps/auth/views/profile.rs [hunk 3]
```diff
@@ -85,25 +91,121 @@ pub async fn profile_update(
...
+		.create(&row)
+		.await
+		.map_err(|e| {
+			error!("Failed to persist verification token: {e}");
+			AppError::Internal("Internal server error".to_string())
+		})?;
+
+	let base_url = std::env::var("PUBLIC_BASE_URL").unwrap_or_default();
+	let url = format!(
+		"{base_url}/auth/verify-email/?user_id={}&token={}",
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/src/apps/auth/views/profile.rs]: Addressed in commit 7be4a3e: the verification link in the email now carries the token in the URL **fragment** (`#user_id=...&token=...`), which is never sent to the server. The dashboard's `/auth/confirm-email` page is responsible for parsing the fragment in JS and POSTing it to the new `/api/auth/verify-email-change/` endpoint as a JSON body, keeping the token out of access logs, proxies, history, and Referer.

File: dashboard/src/apps/auth/views/verify_email.rs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		let stored_arr: [u8; 32] = match stored.try_into() {
+			Ok(a) => a,
+			Err(_) => continue,
+		};
+		// Constant-time equality on fixed-size digests.
+		if stored_arr.ct_eq(&submitted_bytes).unwrap_u8() == 1 {
+			matched = Some(row);
+			break;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/src/apps/auth/views/verify_email.rs]: Addressed in commit 7be4a3e (verify_email_change.rs): the lookup now filters `token_hash = ? AND consumed_at IS NULL AND expires_at > now()` directly at the DB layer and pulls a single row; the application-side scan is gone, so the loop can no longer grow with stale rows.

File: dashboard/src/apps/auth/views/verify_email.rs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+			_ => continue,
+		};
+		let stored_arr: [u8; 32] = match stored.try_into() {
+			Ok(a) => a,
+			Err(_) => continue,
+		};
+		// Constant-time equality on fixed-size digests.
+		if stored_arr.ct_eq(&submitted_bytes).unwrap_u8() == 1 {
+			matched = Some(row);
+			break;
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/src/apps/auth/views/verify_email.rs]: Addressed in commit 7be4a3e: the loop is removed entirely. The handler issues one indexed lookup by `token_hash` (which is UNIQUE) with the freshness predicates pushed into the WHERE clause, so timing depends only on the (uniform) index probe, not on row position.

File: dashboard/migrations/auth/0004_add_email_verification_tokens.rs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+				interleave_in_parent: None,
+				partition: None,
+			},
+			// Composite index to accelerate lookups for a user's unconsumed tokens.
+			Operation::AddConstraint {
+				table: "auth_email_verification_tokens".to_string(),
+				constraint_sql:
+					"CONSTRAINT auth_evt_user_consumed_idx UNIQUE (user_id, token_hash)"
+						.to_string(),
+			},
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/migrations/auth/0004_add_email_verification_tokens.rs]: Addressed in commit 7be4a3e: the redundant `(user_id, token_hash)` UNIQUE is removed (token_hash is already globally unique). It is replaced with a partial index on `user_id` WHERE `consumed_at IS NULL`, which matches the actual per-user pending-token query pattern.

File: dashboard/src/apps/auth/services/mailer.rs [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+		let mut builder = match parsed.scheme.as_str() {
+			"smtps" => AsyncSmtpTransport::<Tokio1Executor>::relay(&parsed.host)
+				.map_err(|e| MailerError::Config(format!("smtps relay: {e}")))?,
+			"smtp" => AsyncSmtpTransport::<Tokio1Executor>::builder_dangerous(&parsed.host),
+			other => {
+				return Err(MailerError::Config(format!(
+					"unsupported scheme: {other}"
+				)));
+			}
+		};
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/src/apps/auth/services/mailer.rs:102]: Addressed in commit 7be4a3e: bare `smtp://` is now rejected with a config error. Production must use `smtps://` (implicit TLS); local dev against MailHog/MailCatcher can opt in explicitly via the new `smtp+insecure://` scheme. The insecure path is no longer the unmarked default.

File: dashboard/src/apps/auth/services/mailer.rs [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+		let (host, port) = match hostport.rsplit_once(':') {
+			Some((h, p)) => {
+				let port: u16 = p
+					.parse()
+					.map_err(|e| MailerError::Config(format!("invalid port: {e}")))?;
+				(h.to_string(), Some(port))
+			}
+			None => (hostport.to_string(), None),
+		};
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/src/apps/auth/services/mailer.rs]: Addressed in commit 7be4a3e: `ParsedSmtpUrl` now wraps `url::Url` (`url` crate, already transitive). Bracketed IPv6 hosts, `@` in passwords, and percent-encoded userinfo all parse correctly. `percent-encoding` is used to decode the credentials.

File: dashboard/src/apps/auth/views/profile.rs [hunk 9]
```diff
@@ -85,25 +91,168 @@ pub async fn profile_update(
...
+		None
+	};
+
+	let resp = serde_json::json!({
+		"profile": ProfileResponse::from(updated),
+		"email_verification": email_verification,
+	});
 	Ok(Response::new(StatusCode::OK)
 		.with_header("Content-Type", "application/json")
 		.with_body(json::to_vec(&resp)?))
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/src/apps/auth/views/profile.rs]: Fixed in commit 6d9b580. `profile_update` now returns `202` with `{ "profile": ..., "email_verification": { "status": "verification_sent" } }` when an email change is requested and the verification email is dispatched. For the no-change and name-only-change cases it returns `200`. When mail delivery fails it returns `200` with `{ "status": "failed" }` so the client can surface a resend affordance. The PR description has been updated to match the actual response shape.

_Generated with [Claude Code](https://claude.com/claude-code)_

File: dashboard/src/apps/auth/views/verify_email_change.rs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+/// Request body for the email-change confirmation endpoint.
+#[derive(Debug, Deserialize)]
+pub struct VerifyEmailChangeRequest {
+	pub user_id: String,
+	pub token: String,
+}
+
+/// Confirm a pending email change.
+#[post("/verify-email-change/", name = "verify_email_change")]
+pub async fn verify_email_change(
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/src/apps/auth/views/verify_email_change.rs:45]: The PR description has been updated in this review cycle. The endpoint is intentionally `POST /auth/verify-email-change/` (not `GET /auth/verify-email/`), and this is the more secure design: the verification email links to a frontend confirmation page (`/auth/confirm-email`) with the token in the URL **fragment** (`#user_id=...&token=...`). Fragments are never transmitted to the server, so the token stays out of access logs and `Referer` headers. The confirmation page reads the fragment in JavaScript and POSTs the token as a JSON body to `POST /api/auth/verify-email-change/`. The PR description now accurately reflects this flow.

_Generated with [Claude Code](https://claude.com/claude-code)_

File: dashboard/src/apps/auth/views/profile.rs [hunk 11]
```diff
@@ -85,25 +91,168 @@ pub async fn profile_update(
...
+	// A single-row-at-a-time delete loop keeps us compatible with the ORM's
+	// primary-key-based delete API.
+	let prior = EmailVerificationToken::objects()
+		.filter(
+			EmailVerificationToken::field_user_id(),
+			FilterOperator::Eq,
+			FilterValue::String(user.id.to_string()),
+		)
+		.all()
+		.await
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/src/apps/auth/views/profile.rs:210]: Fixed in commit 6d9b580. The stale-token cleanup query in `issue_email_verification` now includes a `consumed_at IS NULL` filter at the DB layer (via `FilterOperator::IsNull`), which matches the partial index predicate (`WHERE consumed_at IS NULL`) added by migration 0004. Consumed rows are no longer fetched, so the partial index is engaged, the result set is bounded to genuinely pending tokens, and the in-application `consumed_at.is_some()` skip is removed.

_Generated with [Claude Code](https://claude.com/claude-code)_

File: dashboard/src/apps/auth/tests/unit/test_email_verification.rs [hunk 12]
```diff
@@ -0,0 +1,90 @@
+//! Unit tests for the email-change verification flow.
+//!
+//! These tests cover the security-critical building blocks that do NOT
+//! require a database: token generation properties, SHA-256 hashing,
+//! constant-time hash comparison, and the [`NullEmailSender`]. Full
+//! request-level happy/expired/consumed/invalid-token coverage requires
+//! a PostgreSQL TestContainer and is tracked as an e2e follow-up.
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/src/apps/auth/tests/unit/test_email_verification.rs]: Addressed in commit 6d9b580. Four new unit tests cover the edge cases the handler checks before acting:

- `test_expired_token_is_detected` — verifies that `expires_at` in the past correctly signals expiry (matches the DB `expires_at > now()` filter).
- `test_consumed_token_is_detected` — verifies that `consumed_at.is_some()` correctly signals a consumed token (matches the DB `consumed_at IS NULL` filter).
- `test_user_id_mismatch_is_detected` — verifies that a different `user_id` does not match the token owner (matches the post-lookup `token.user_id != claimed_user_id` check).
- `test_valid_token_passes_all_checks` — confirms that a well-formed token passes all three conditions.

Full DB-backed request-level tests (reissue-deletes-prior, uniqueness conflict on email apply) require the Postgres testcontainer and are tracked as a follow-up.

_Generated with [Claude Code](https://claude.com/claude-code)_

File: dashboard/src/apps/auth/tests/unit/test_email_verification.rs [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+		let d2: [u8; 32] = h2.finalize().into();
+
+		// Assert
+		assert_eq!(d1, d2);
+	}
+
+	/// Constant-time compare returns 1 for equal digests and 0 for
+	/// different digests (invalid-token rejection path).
+	#[rstest]
+	fn test_constant_time_eq_rejects_mismatched_digest() {
```
[the remark this answers is not in the document]
Comment (kent8192) [reply dashboard/src/apps/auth/tests/unit/test_email_verification.rs]: Addressed in commit 6d9b580. The `EmailVerificationToken` model doc has been updated to explain why the DB equality lookup is timing-safe without application-level `ConstantTimeEq`:

> Because the lookup is by exact hash match (not by scanning a candidate list), timing depends on the uniform index lookup rather than on token position in a list. A constant-time comparison at the application layer would add no additional security here: the token hash is globally unique (UNIQUE constraint), so there is at most one candidate row, and the timing of a DB index lookup does not reveal anything useful to an attacker. Application-level constant-time comparison would only be relevant if we were scanning multiple rows in memory — which we do not.

The PR description has also been updated to replace the inaccurate `subtle::ConstantTimeEq` claim with the accurate description of the DB lookup approach.

_Generated with [Claude Code](https://claude.com/claude-code)_
